docs/problem_format/cpp_psetting_templates.md

```diff
@@ -0,0 +1,63 @@
+- `std::string readLine(char min_char = 0, char max_char = 127)` returns the next line in the input stream. Specifically, it reads until it encounters a `\n`, and discards it (the newline is not part of the returned string). `min_char` and `max_char` are the same as for `readToken`. If `readLine` encounters an EOF, it fails.
+- `long long readInt(long long lo, long long hi)` parses the next token as an integer. It aborts on overflow, malformed integers, and if the resultant integer is not in the range [lo, hi], inclusive. Leading zeroes and `-0` are not accepted.
+- `long double readFloat(long double lo, long double hi, long double eps = 1e-9)` parses the next token as a float. It aborts on overflow, malformed floats, and if the resultant float is not in the range [lo, hi], inclusive, using the provided epsilon to perform the comparison. Scientific notation and NaNs are not accepted, nor are leading zeroes. `-0` is allowed. Trailing zeroes are also permitted.
+- `std::vector<T> readIntArray(size_t N, long long lo, long long hi)` parses the next space-separated N integers into an array, and then reads a final newline. It must be given a template argument, which is the type of the array elements. For example, `readIntArray<int>(5, 1, 10)` reads five space-separated integers into a `std::vector<int>`, where each integer is in the range [1, 10], inclusive.
+
+A small caveat: `readToken` and `readLine` will throw if the string exceeds 10 million characters.
+
+`readFloat()` will likely be of no use for many validators, and can be safely deleted. Similarly, `readIntArray` can be deleted if unneeded.
+
+## Checkers/Interactors
+
+The next pair of templates are for checkers/interactors. The difference is the type of whitespace handling: the identical checker/interactor expects whitespace to match exactly. The standard checker/interactor handles whitespace like the `standard` checker.
+
+The checkers and interactors are designed for the `coci` bridged checker/interactor type. However, updating the codes used and the order of command line parameters to work with other types should not be challenging.
+
+Both files can be used for either checkers/interactors, with the following caveat: interactors MUST close `stdout` BEFORE calling `readEOF()`, so that the user process can terminate in case it _also_ expects an EOF. Checker stdout is used for feedback displayed to the user, and as such `stdout` should not be closed in this case. Validators also do not need to worry about this - only interactors do, and they should only call `readEOF()` once they have finished communicating with the user, to clean up and assert that the user didn't send any trailing data.
+
+The general format of the checkers/interactors are the same as the validator, with a few changes:
+
+- `readSpace(), readNewLine(), readEOF()`: Under the identical checker, these return Presentation Error if the check fails. Under the standard checker, these return WA.
```

Comment:
I don't remember how the standard checker works off the top of my head anymore: does it treat all whitespace interchangeably (i.e. you can replace a space with a newline and vice versa) or does it only strip out empty lines and ignore extra space?

In either case, should we consider adding methods that read any whitespace (since `readSpace()` and `readNewLine()` will throw if they read a newline or space, respectively), and also one that consumes all whitespace?

Reply:
The standard checker:
- Consumes all whitespace at start.
- Treats newlines and non-newline WS separately.
- Discards lines with no non-WS chars.

The behaviour of the methods is:
- `readSpace` expects to find non-newline WS before the next token
- `readNewLine` expects to find newline WS before the next token
- `readEOF` expects to find no more tokens

I am against adding methods that do weird shit to whitespace because it doesn't really match how standard checker works. The way the methods are setup right now, you are supposed to call the methods in a way that describes the canonical input format (single space-separated, no blank lines, newline before EOF).

"Consume all whitespace" or "Consume some whitespace" are both not part of the normal standard checker algorithm (except at start and end ofc).

Or put another way, I'd probably reject a checker that used these methods if I wrote them.

Checkers that want very fine-grained control on WS and stuff should either use the identical template and do custom processing or write some code from scratch. I don't want to complicate these templates for the 1%, I want them to be simple enough for the 99% instead. I agree that they don't cover all cases

Reply:
I think "consume all whitespace" is part of the normal standard checker routine:
https://github.com/DMOJ/judge-server/blob/master/dmoj/checkers/_checker.c#L58-L59

That being said, I am ok with not implementing this, as long as it's noted that this differs from the behaviour of the `standard` checker in the judge repo.

Reply:
`skip_spaces` moves the cursor until the next token and reports the type of WS found: either "no lines" or "some lines" (or no WS). It then checks that the same type was seen by the other buffer. This corresponds to `readSpace` and `readNewLine` respectively

Reply:
It seems the methods were just terribly documented, so I confused you - overhauled the docs